fuse_variables/include/fuse_variables/pinhole_camera.h

```diff
@@ -0,0 +1,184 @@
+/*
+ * Software License Agreement (BSD License)
+ *
+ *  Author:    Oscar Mendez
+ *  Created:   11.13.2023
+ *
+ *  Copyright (c) 2023, Locus Robotics
+ *  All rights reserved.
+ *
+ *  Redistribution and use in source and binary forms, with or without
+ *  modification, are permitted provided that the following conditions
+ *  are met:
+ *
+ *   * Redistributions of source code must retain the above copyright
+ *     notice, this list of conditions and the following disclaimer.
+ *   * Redistributions in binary form must reproduce the above
+ *     copyright notice, this list of conditions and the following
+ *     disclaimer in the documentation and/or other materials provided
+ *     with the distribution.
+ *   * Neither the name of the copyright holder nor the names of its
+ *     contributors may be used to endorse or promote products derived
+ *     from this software without specific prior written permission.
+ *
+ *  THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
+ *  "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
+ *  LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS
+ *  FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE
+ *  COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT,
+ *  INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING,
+ *  BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES;
+ *  LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+ *  CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+ *  LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN
+ *  ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
+ *  POSSIBILITY OF SUCH DAMAGE.
+ */
+#ifndef FUSE_VARIABLES_PINHOLE_CAMERA_H
+#define FUSE_VARIABLES_PINHOLE_CAMERA_H
+
+#include <fuse_core/fuse_macros.h>
+#include <fuse_core/serialization.h>
+#include <fuse_core/uuid.h>
+#include <fuse_variables/fixed_size_variable.h>
+
+#include <boost/serialization/access.hpp>
+#include <boost/serialization/base_object.hpp>
+#include <boost/serialization/export.hpp>
+
+#include <ostream>
+
+namespace fuse_variables
+{
+/**
+ * @brief Variable representing intrinsic parameters of a camera.
+ *
+ * The UUID of this class is constant after
+ * construction and dependent on a user input database id. As such, the database id cannot be altered after
+ * construction.
+ */
+class PinholeCamera : public FixedSizeVariable<4>
+{
+public:
+  FUSE_VARIABLE_DEFINITIONS(PinholeCamera);
+
+  /**
+   * @brief Can be used to directly index variables in the data array
+   */
+  enum : size_t
+  {
+    FX = 0,
+    FY = 1,
+    CX = 2,
+    CY = 3
+  };
+
+  /**
+   * @brief Default constructor
+   */
+  PinholeCamera() = default;
+
+  /**
+   * @brief Construct a pinhole camera variable given a camera id
+   *
+   * @param[in] camera_id  The id associated to a camera
+   */
+  explicit PinholeCamera(const uint64_t& camera_id);
```

Comment:
For positions/orientations, I use a UUID as a mechanism for differentiating between poses from different robots. My reasoning goes like this: Most robots have some sort of name string, and a UUID can be easily generated directly from the name. That makes distributed systems easier. Each robot can generate an ID number for itself without having to coordinate with some central system. And if the user really just wants an index, you can cram an integer inside a UUID container _without too much hassle_.

If I imagine a multi-robot system where each robot has multiple cameras, how do I coordinate assigning each camera a unique camera index correctly? Whereas I could generate a UUID from the robot name + camera name strings easily. I might even be able to pass in the camera image header frame name directly.

For landmarks, I'll be honest. I didn't really have a plan. Landmarks don't seem tied to a specific piece of hardware in the same way. If the landmarks are all fixed, then they probably come from a central database anyway, and assigning ID numbers is probably easy. If the landmarks come from some sort of ARTag, then they have an ID built into them. But if you are trying to build up a database of image feature landmarks from multiple robots, the ID field may get cumbersome.

---

Unrelated to ID numbers. A robot general has multiple sensors. Those sensors are mounted at some fixed location on the robot. Extrinsic calibration if you will. If I have a variable that represents the pose of the robot in the global frame, and that robot has a camera mounted with some non-Identity transform from the main robot frame, how do we include observations from the camera in the same graph as wheel encoder data in the robot frame? Would we need a "MountedCamera" variable that includes the base->camera transform? Or would that really get handled in the constraint equation? This is a topic on the ToDo list I keep in my head. I just haven't needed to tackle it yet.

---

And so ends this edition of "Steve's Random Thoughts".